src/vanish_subscriber.rs

```diff
@@ -0,0 +1,172 @@
+        match self.con.set(key, value.clone()).await {
+            Ok(()) => Ok(()),
+            Err(_) => self.con.set(key, value).await,
+        }
+    }
+
+    async fn xread_options(
+        &mut self,
+        keys: &[&str],
+        ids: &[String],
+        opts: &StreamReadOptions,
+    ) -> Result<StreamReadReply, RedisError> {
+        match self.con.xread_options(keys, ids, opts).await {
+            Ok(reply) => Ok(reply),
+            Err(_) => self.con.xread_options(keys, ids, opts).await,
+        }
+    }
+}
+
+pub fn start_vanish_subscriber<T: RedisClientTrait, U: RepoTrait>(
+    tracker: TaskTracker,
+    redis_client: T,
+    repo: Arc<U>,
+    cancellation_token: CancellationToken,
+) {
+    tracker.spawn(async move {
+        let (mut con, mut last_id) = match get_connection_and_last_id(redis_client).await {
+            Ok(result) => result,
+            Err(e) => {
+                error!("Failed to get Redis connection: {}", e);
+                return;
+            }
+        };
+
+        let opts = StreamReadOptions::default().block(BLOCK_MILLIS);
+
+        info!("Starting from last id processed: {}", last_id);
+
+        loop {
+            tokio::select! {
+                _ = cancellation_token.cancelled() => {
+                    break;
+                }
+
+                result = async {
+                    let reply: StreamReadReply = con
+                        .xread_options(&[VANISH_STREAM_KEY], &[last_id.clone()], &opts)
+                        .await?;
+
+                    for StreamKey { ids, .. } in reply.keys {
+                        for stream_id in ids {
+                            if stream_id.id == last_id {
+                                continue;
+                            }
+
+                            let Some(value) = stream_id.map.get("pubkey") else {
+                                error!("Vanish request doesn't have a public key");
+                                continue;
+                            };
+
+                            let public_key = match value {
+                                redis::Value::BulkString(bytes) => {
+                                    let public_key_string = String::from_utf8(bytes.clone())?;
+
+                                    PublicKey::from_hex(public_key_string)?
+                                }
+                                _ => {
+                                    error!("Vanish request public key is not a bulk string");
+                                    continue;
+                                }
+                            };
+
+                            if let Err(e) = repo.remove_pubkey(&public_key).await {
+                                error!("Failed to remove public key: {}", e);
+                            }
+
+                            info!("Removed public key {} from vanish request", public_key.to_hex());
+
+                            last_id = stream_id.id.clone();
```

Comment:
Does this get stored somewhere other than memory?

Reply:
Oops, thanks!